### PR TITLE
issue-72-Dropdown

### LINKED ISSUE #72: Dropdown
## General info
Create a small dropdown button that is used to set some state, e.g. if state changes via the dropdown filtering on the page changes.

## Note
Similar to `DropdownInput` component, can pretty much copy its code.

image

### PR DESCRIPTION
**Pull request by ErikaKK**
## Change Summary
**[Briefly summarise the changes that you made. Just high-level stuff]**
image
Add resizable dropdown icon component `MarketDropdownIcon`. Add `MarketDropdown` component.
 * Dropdown component in bidders side market page with a list of options.
   * @param props - The properties for the DropdownInput component.
   * @param props.value - Value of the selected option, tracked by `react.setState()`.
   * @param props.options - List of possible option for that category, displayed in dropdown.
   * @param props.onChange - Called when the selected value changes, should update `value`.
   * @param props.label - Label for the field, display "All ...".
### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [ ] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [ ] The change has documentation

## Other Information
**[Is there anything in particular in the review that I should be aware of?]**
Should I set the maximum height of the dropdown menu?

# Related issue

- Resolve #72

### CODE REVIEW

File: client/src/components/ui/icons.tsx [hunk 1]
```diff
@@ -142,3 +142,27 @@ const MeIcon: React.FC<BottomNavIconProps> = ({ className, ...props }) => {
...
+        stroke="currentColor"
+        strokeWidth={2}
+        strokeLinecap="round"
+        strokeLinejoin="round"
+      />
+    </svg>
+  );
+};
+
+export { MarketDropdownIcon };
```
Comment (SodaVolcano) [inline client/src/components/ui/icons.tsx]: Unnecessary, dropdown icon already exist. Also following the style guide please use `export function SomeIcon(...) {...}` instead of `const SomeIcon: ...`

File: client/src/components/ui/market-dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+const valueStyle =
+  "body  h-full w-full resize-none bg-transparent text-penni-text-regular-light-mode caret-penni-main focus:outline-none";
+
+// Generate unique ID for each component, used for label htmlFor attribute
+const uniqueId = () => `${Date.now()}-${Math.random()}`;
+
+type HTMLEventTargetElement =
+  | HTMLInputElement
+  | HTMLTextAreaElement
+  | HTMLSelectElement;
```
Comment (SodaVolcano) [inline client/src/components/ui/market-dropdown.tsx]: Probably not needed here, you should be able to just use `React.ChangeEvent` without `<...>`?

File: client/src/components/ui/market-dropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+interface DropdownInputProps {
+  value: string;
+  onChange: (e: React.ChangeEvent<HTMLEventTargetElement>) => void;
```
Comment (SodaVolcano) [inline client/src/components/ui/market-dropdown.tsx]: Try changing this to `(e: React.ChangeEvent)`

File: client/src/components/ui/market-dropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  label: string;
+  options: string[];
+}
+
+interface DropdownMenuProps {
+  menuId: string;
+  options: string[];
+  label: string;
+  onChange: (e: React.ChangeEvent<HTMLEventTargetElement>) => void;
```
Comment (SodaVolcano) [inline client/src/components/ui/market-dropdown.tsx]: Same here

File: client/src/components/ui/market-dropdown.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+              }
+            >
+              {option}
+            </li>
+          ))}
+        </ul>
+      </div>
+    </div>
+  );
+}
```
Comment (SodaVolcano) [inline client/src/components/ui/market-dropdown.tsx]: Looks mostly copied from `Input`, consider moving this to its own `dropdown-menu.tsx` file and pass in `[label, ...options]` instead of storing it inside the component. The size can maybe be changed via a `width` parameter, though maybe it should take up full width and the parent element can contain its size. i.e. is it possible to size `DropdownMenu` to the width of `MarketDropdown`? That way the code in `DropdownMenu` stays the same so we don't need a new parameter.

File: client/src/components/ui/market-dropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                {label}
+              </label>
+            ) : (
+              <span className={`${valueStyle} text-left`}>{value}</span>
+            )}
+
+            <div className="flex w-2.5 items-center">
+              <MarketDropdownIcon className="w-full text-penni-text-regular-light-mode" />
+            </div>
+          </div>
```
Comment (SodaVolcano) [inline client/src/components/ui/market-dropdown.tsx]: Hm.. Perhaps both `DropdownMenu` and this component (rename to `DropdownButton`) should be moved to `dropdown.tsx`, where all elements inside `<button>` (highlighted here) in both `market-dropdown.tsx` and similar lines in `input.tsx` should be passed as a `{children}` prop? We can keep `w-full` and encase the button inside a div, and change the width of that div to size the button (so we don't have to change the code).

**Review by SodaVolcano — CHANGES_REQUESTED** [review]
Maybe the code can be refactored so we don't have duplicates in `input.tsx` and `market-dropdown.tsx`. To summarise:
- Create a new component file `dropdown.tsx` and move `DropdownMenu`, all the style strings and interfaces at the start of the file, and the outer part of `MarketDropdown` excluding elements inside `<button>` into this file (rename this element to `DropdownButton`)
    - This way, `DropdownButton`'s sole role is to spawn a menu, what the button looks like can be customised
    - If the `DropdownMenu` have different width from the button then we can pass `DropdownButton` a `menuWidth` prop
- Keep the same component name (i.e. keep `DropdownInput` and `MarketDropdown`) in the original files, but this time they use `DropdownButton` and pass in elements inside `<button>` like

```js
export function MarketDropdown(...) {
  ...
  return (
    <div className="w-36 ...">      // Use div to size the elements inside
      <DropdownButton options={[label, ...options]} ...>
        {/* element that differs from DropdownInput here... */}
      </DropdownButton>
    </div>
  )
}
```

Replace code for `DropdownInput` as well.

File: client/src/components/ui/icons.tsx (hunk 1 above)
Comment (SodaVolcano) [reply client/src/components/ui/icons.tsx]: btw feel free to edit the `DropdownIcon`, but keep in mind it have to work with usage in `DropdownInput` as well

File: client/src/components/ui/dropdown.tsx [hunk 7]
```diff
@@ -1,16 +1,13 @@
...
 
 // Change background colour and show border when onFocus
 // Uses border-opacity because adding border-2 shifts the content inside div
 const selectedStyle =
   "border-opacity-100 border-penni-grey-border-light-mode bg-penni-main-shade2";
 const deselectedStyle =
   "border-penni-grey-border-light-mode bg-penni-background-light-mode";
-// Padding changes when onFocus as label size changes and children is rendered
-const expandedStyle = "px-4 justify-center";
-const collapsedStyle = "py-4 pl-4 pr-3";
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Why was this removed? `label` should still be optional so these styles are needed I think

File: client/src/components/ui/dropdown.tsx [hunk 8]
```diff
@@ -28,33 +25,69 @@ type HTMLEventTargetElement =
...
   label: string;
   options: string[];
 }
 
+interface DropdownButtonProps {
+  children: React.ReactNode;
+  id: string;
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+  style: string;
+  caretWidth: number; //px
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Is there a reason for `caretWidth` to require different width? Also caret shouldn't be displayed at all if the input is a dropdown?

File: client/src/components/ui/dropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+  options: string[];
+  onChange: (e: React.ChangeEvent<HTMLEventTargetElement>) => void;
+}
+
+export function DropdownButton({
+  id,
+  children,
+  onClick,
+  style,
+  caretWidth,
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: what's the purpose of `caretWidth`?

File: client/src/components/ui/dropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+const valueStyle =
+  "body  h-full w-full resize-none bg-transparent text-penni-text-regular-light-mode caret-penni-main focus:outline-none";
+
+// Generate unique ID for each component, used for label htmlFor attribute
+const uniqueId = () => `${Date.now()}-${Math.random()}`;
+
+type HTMLEventTargetElement =
+  | HTMLInputElement
+  | HTMLTextAreaElement
+  | HTMLSelectElement;
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Remove this, not needed

File: client/src/components/ui/dropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+interface MarketDropdownProps {
+  value: string;
+  onChange: (e: React.ChangeEvent<HTMLEventTargetElement>) => void;
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Just use `(e: React.ChangeEvent)`

File: client/src/components/ui/dropdown.tsx [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+  id: string;
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+  style: string;
+  caretWidth: number; //px
+}
+
+interface DropdownMenuProps {
+  menuId: string;
+  options: string[];
+  onChange: (e: React.ChangeEvent<HTMLEventTargetElement>) => void;
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Same here, use `(e: React.ChangeEvent)`

File: client/src/components/ui/dropdown.tsx [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        )}
+      </DropdownButton>
+
+      {isExpanded && (
+        <DropdownMenu
+          menuId={menuId}
+          options={optionList}
+          onChange={handleOnChange}
+        />
+      )}
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Why is `isExpanded` logic here? Should be in `DropdownButton`, all that's needed in `MarketDropdown` is `<DropdownButton>{elements here...}</DropdownButton>`

File: client/src/components/ui/dropdown.tsx [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+   * />
+   */
+export function MarketDropdown({
+  value,
+  options,
+  onChange,
+  label,
+}: MarketDropdownProps) {
+  const [menuId] = useState(uniqueId());
+  const [isExpanded, setExpanded] = useState(false); // isSelect in text inputs
```
Comment (SodaVolcano) [inline client/src/components/ui/dropdown.tsx]: Belongs in `<DropdownButton>`, also remove the comment since this is not in `inputs.tsx` anymore

File: client/src/components/ui/inputs.tsx [hunk 15]
```diff
@@ -244,39 +245,39 @@ export function ParagraphInput({
...
+//               }}
+//             >
+//               {option}
+//             </button>
+//           ))}
+//         </div>
+//       </div>
+//     </div>
+//   );
+// }
```
Comment (SodaVolcano) [inline client/src/components/ui/inputs.tsx]: Remove, don't leave chunks of commented out code

File: client/src/components/ui/inputs.tsx [hunk 16]
```diff
@@ -313,41 +314,36 @@ export function DropdownInput({
...
+            </label>
+          )}
+          {value !== "" && (
+            <span className={`${valueStyle} text-left`}>{value}</span>
+          )}
+        </div>
+      </DropdownButton>
 
       {isExpanded && (
         <DropdownMenu
```
Comment (SodaVolcano) [inline client/src/components/ui/inputs.tsx]: Remove, logic belongs in `<dropdownButton>`

**Review by SodaVolcano — CHANGES_REQUESTED** [review]
Still need quite a few refactoring. See comments from previous review

**Comment by SodaVolcano** [comment]
Also could you do `git merge main` to pull the commits from `main`? And add your dropdown to the `component-showcase` page :+1:

File: client/src/components/ui/dropdown.tsx (hunk 7 above)
Comment (ErikaKK) [reply client/src/components/ui/dropdown.tsx]: These two variable are used to define the `containerStyle` in `inputs.tsx`, not in `dropdown.tsx`.

File: client/src/components/ui/dropdown.tsx (hunk 8 above)
Comment (ErikaKK) [reply client/src/components/ui/dropdown.tsx]: Because the caret size is different in MarketDropdown

File: client/src/components/ui/dropdown.tsx (hunk 13 above)
Comment (ErikaKK) [reply client/src/components/ui/dropdown.tsx:177]: Ok I know what you mean now I'll move these to <DropdownButton>

File: client/src/components/ui/dropdown.tsx (hunk 10 above)
Comment (ErikaKK) [reply client/src/components/ui/dropdown.tsx]: If I remove these types, I'll  have to make changes in `inputs.tsx` accordingly

File: client/src/components/ui/inputs.tsx (hunk 16 above)
Comment (ErikaKK) [reply client/src/components/ui/inputs.tsx]: `DropdownInput` component:
image
`MarketDropdown` component:
image

They have slightly different logic because the `containerStyle` of the `DropdownButton` is different based on `isExpended`. It is hard to move the `isExpended` part into `DropdownButton`

**Comment by yunho7687** [comment]
@SodaVolcano @ErikaKK good work! when it is good to merge please let me know

File: client/src/components/ui/inputs.tsx (hunk 16 above)
Comment (SodaVolcano) [reply client/src/components/ui/inputs.tsx]: Could you highlight what style is affected by moving `isExpanded` into the button component? I can have a look if needed

Comment (SodaVolcano) [reply client/src/components/ui/inputs.tsx]: Also, please add `onBlur` to `DropdownButton` that set `isExpanded` to false so the dropdown menu closes when the user clicks away. I forgot to add this in when making the input components

File: client/src/components/ui/dropdown.tsx (hunk 8 above)
Comment (SodaVolcano) [reply client/src/components/ui/dropdown.tsx]: Caret is the vertical bar that appears when you click on a text input (i.e. the cursor), dropdown is a button so the caret shouldn't be displayed at all??

**Review by yunho7687 — APPROVED** [review]

File: client/src/components/ui/dropdown.tsx (hunk 8 above)
Comment (ErikaKK) [reply client/src/components/ui/dropdown.tsx]: I mean the dropdownIcon

**Review by SodaVolcano — APPROVED** [review]
